### PR TITLE
Show service logs for user services

### PR DESCRIPTION
**Pull request by t1m0thyj**
# Services: Show logs for user units

The Services page now shows journal logs for user systemd services.

Screen Shot 2023-01-10 at 14 54 35

---------------------------
Removes the limitation mentioned in https://github.com/cockpit-project/cockpit/commit/154da47d947eb39ebf2bed4cf6e1ecfc1623a25e:
> * Hide service logs for user services. These are available but the
LogsPanel component will need some extension for supporting it.

cockpit-user-service-logs

### CODE REVIEW

**Review by KKoukiou — COMMENTED** [review]
@t1m0thyj thanks a lot for the contribution. This will need some test to get in. The correct place to add the test IMO is testLogs, we can create a user test service and expect some mock logs, exactly how we do that for the system one. Are you interested in doing that?

If not I will send a commit on top of your PR. Let me know.

File: pkg/systemd/logsHelpers.js [hunk 1]
```diff
@@ -58,6 +58,13 @@ export const getGrepFiltersFromOptions = ({ options }) => {
             match.push(...['_SYSTEMD_UNIT=' + s, "+", "COREDUMP_UNIT=" + s, "+", "UNIT=" + s]);
         });
         full_grep += "service:" + options.service + " ";
+    } else if (options["user-service"]) {
+        options["user-service"].split(",").forEach(s => {
+            if (!s.endsWith(".service"))
+                s = s + ".service";
+            match.push(...['_SYSTEMD_USER_UNIT=' + s, "+", "COREDUMP_USER_UNIT=" + s, "+", "USER_UNIT=" + s]);
+        });
+        full_grep += "user-service:" + options["user-service"] + " ";
```
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:67]: These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221214-134315-b7b4124e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by t1m0thyj** [comment]
> @t1m0thyj thanks a lot for the contribution. This will need some test to get in. The correct place to add the test IMO is testLogs, we can create a user test service and expect some mock logs, exactly how we do that for the system one. Are you interested in doing that?
> 
> If not I will send a commit on top of your PR. Let me know.

@KKoukiou Thanks for the quick review 🙂 I'll take a stab at adding the necessary tests and let you know if I need help.

**Comment by t1m0thyj** [comment]
I've added a new test `testLogsUser` which seems to have broken other tests and I'm not sure why, would appreciate if you can take a look.

**Comment by jelly** [comment]
> I've added a new test `testLogsUser` which seems to have broken other tests and I'm not sure why, would appreciate if you could please take a look.

The TestAccounts.testBasic failure is indeed unrelated, I have seen this in other Pull Requests as well.

**Comment by KKoukiou** [comment]
@t1m0thyj the testing-farm failures are because you PR is not rebased. If you find time please rebase on top of origin/main.

File: pkg/systemd/logDetails.jsx [hunk 2]
```diff
@@ -55,7 +57,7 @@ const LogDetails = ({ entry }) => {
                     </CardHeaderMain>
                     { service &&
                         <CardActions>
-                            <Button variant="link" onClick={() => cockpit.jump("/system/services#/" + service) }>
+                            <Button variant="link" onClick={() => cockpit.jump("/system/services#/" + service + (is_user ? "?owner=user" : "")) }>
```
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221216-124213-5d867205-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 1 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:67]: These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221216-124213-5d867205-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by t1m0thyj** [comment]
> @t1m0thyj the testing-farm failures are because you PR is not rebased. If you find time please rebase on top of origin/main.

Will do when I find some time 👍 Do I also need to add unit tests to cover the lines that @cockpituous has commented on? It seems like the tests I added so far are system tests that don't affect coverage.

**Comment by KKoukiou** [comment]
@t1m0thyj the new tests are failing. You can view the links from any of the distros that we test. I am happy to test that tomorrow for you as well. Example failure: https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221219-171258-00d16b76-fedora-37/log.html#326

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221219-171301-00d16b76-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 1 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:67]: These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221219-171301-00d16b76-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by t1m0thyj** [comment]
> @t1m0thyj the new tests are failing. You can view the links from any of the distros that we test. I am happy to test that tomorrow for you as well. Example failure: https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221219-171258-00d16b76-fedora-37/log.html#326

Hopefully tests are fixed now 🙂 If you don't mind rerunning tests I would appreciate it, as setting up a test environment in a clean VM seems like it would require a significant amount of effort.

**Comment by KKoukiou** [comment]
@t1m0thyj I sent a fix up commit for the tests, now its passing locally at least :) waiting for the ci to see. thanks for the feature btw.

**Review by KKoukiou — DISMISSED** [review]

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221220-191931-16e0a784-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js [hunk 3]
```diff
@@ -58,6 +58,13 @@ export const getGrepFiltersFromOptions = ({ options }) => {
             match.push(...['_SYSTEMD_UNIT=' + s, "+", "COREDUMP_UNIT=" + s, "+", "UNIT=" + s]);
         });
         full_grep += "service:" + options.service + " ";
+    } else if (options["user-service"]) {
```
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221220-191931-16e0a784-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js [hunk 4]
```diff
@@ -58,6 +58,13 @@
...
+        options["user-service"].split(",").forEach(s => {
+            if (!s.endsWith(".service"))
+                s = s + ".service";
```
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221220-191931-16e0a784-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by KKoukiou** [comment]
Some failures that look related.
Retrying once more before checking the tests manually.
https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20221220-191917-16e0a784-debian-testing/log.html#326

**Comment by KKoukiou** [comment]
@jelly can you please remind me what's with this issue with missing journal messages in our logs page tests? I see this flaking here, and I recall you have dealt with this recently.

**Comment by jelly** [comment]
> @jelly can you please remind me what's with this issue with missing journal messages in our logs page tests? I see this flaking here, and I recall you have dealt with this recently.

We had something similiarish in [testLogs]( https://github.com/cockpit-project/cockpit/issues/17768) and we have https://github.com/cockpit-project/cockpit/pull/17999 to fix it. So let's review this PR and see if we need a similar fxi

**Comment by jelly** [comment]
From the logs:

```
Jan 04 20:10:48 debian sh[2437]: START
Jan 04 20:10:48 debian systemd[2287]: Started Test Service.
```

**Comment by jelly** [comment]
Please remove `test/verify/.check-system-services.swp` from the PR it shouldn't be included.

**Comment by jelly** [comment]
Created https://github.com/systemd/systemd/issues/25932 for the issue, I think we should exclude checking for "START" in systemd 252 as it's a known issue.

**Comment by KKoukiou** [comment]
> Created [systemd/systemd#25932](https://github.com/systemd/systemd/issues/25932) for the issue, I think we should exclude checking for "START" in systemd 252 as it's a known issue.

But how come it does not consistently appear in our current tests on main @jelly?

**Comment by jelly** [comment]
> > Created [systemd/systemd#25932](https://github.com/systemd/systemd/issues/25932) for the issue, I think we should exclude checking for "START" in systemd 252 as it's a known issue.
> 
> But how come it does not consistently appear in our current tests on main @jelly?

It only fails in Arch / Debian-testing right? Both those images have systemd 252 while the others have systemd < 252.

**Comment by jelly** [comment]
Turns out the log order is not guaranteed, so we should fix our testing method https://github.com/systemd/systemd/issues/25932#issuecomment-1372100404

**Comment by KKoukiou** [comment]
Reposted with a fix for the systemd issue @jelly pointed out.

**Comment by t1m0thyj** [comment]
I see that a number of tests are failing now, let me know if there are any changes I should make to the PR 🙂

**Comment by KKoukiou** [comment]
> I see that a number of tests are failing now, let me know if there are any changes I should make to the PR slightly_smiling_face

Sorry this is my fault. Debugging as well locally.

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-091646-f92cc1a9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-091646-f92cc1a9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-091646-f92cc1a9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-114858-7e8da5d0-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-114858-7e8da5d0-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-114858-7e8da5d0-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-142642-9203b7d7-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-142642-9203b7d7-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-142642-9203b7d7-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-154907-d3f2ab25-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-154907-d3f2ab25-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230110-154907-d3f2ab25-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230111-070139-a66d2c08-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 1 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:67]: These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230111-070139-a66d2c08-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230111-081109-50ec7354-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230111-081109-50ec7354-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230111-081109-50ec7354-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by KKoukiou — DISMISSED** [review]

**Review by jelly — APPROVED** [review]

File: test/verify/check-system-services [hunk 5]
```diff
@@ -154,6 +155,7 @@ WantedBy=default.target
 
 trap "echo STOP" 0
 
+python -c "import systemd.daemon, time; systemd.daemon.notify('READY=1'); time.sleep(5)"
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply test/verify/check-system-services]: This is outdates.

File: test/verify/check-system-services [hunk 6]
```diff
@@ -641,11 +638,23 @@ ExecStart=/bin/awk -f /tmp/mem-hog{params}.awk
             b.wait_js_cond(f"window.location.hash === '#/{suffix}'")
 
     def testLogs(self):
+        self._testLogs(False)
+
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply test/verify/check-system-services]: This is oudated.

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230116-112945-90f9e191-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230116-112945-90f9e191-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230116-112945-90f9e191-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
This rhel-9-2 failure is very stubborn.. 

```
Jan 19 11:26:09 rhel-9-2-127-0-0-2-2201 sh[2722]: START
Jan 19 11:26:09 rhel-9-2-127-0-0-2-2201 systemd[2270]: Started Test Service.
Jan 19 11:26:10 rhel-9-2-127-0-0-2-2201 PackageKit[2315]: resolve transaction /1_ededceaa from uid 0 finished with failed after 801ms
Jan 19 11:26:10 rhel-9-2-127-0-0-2-2201 PackageKit[2315]: get-updates transaction /2_accddeaa from uid 0 finished with failed after 374ms
Jan 19 11:26:10 rhel-9-2-127-0-0-2-2201 systemd[1]: Started RHSM dbus service.
Jan 19 11:26:14 rhel-9-2-127-0-0-2-2201 sh[2722]: WORKING
Jan 19 11:26:19 rhel-9-2-127-0-0-2-2201 sh[2722]: WORKING
```

**Comment by martinpitt** [comment]
That failure is *exactly* what https://github.com/systemd/systemd/issues/25932 discussed. The [commit message](https://github.com/cockpit-project/cockpit/pull/18032/commits/90f9e1915a1d9fc151fca1206b10f2d50dd9d657) claims it applies that solution (`Type=notify`), but does not actually do it. I suggest we fix that in the existing test case first, and then apply it to the user unit/journal in this PR as well.

File: test/verify/check-system-services [hunk 7]
```diff
@@ -45,6 +45,7 @@ class TestServices(MachineCase):
         user_post_restore = "su - admin -c 'export XDG_RUNTIME_DIR=/run/user/$(id -u admin); systemctl --user daemon-reload; systemctl --user reset-failed'"
         self.restore_dir("/etc/systemd/user", user_post_restore, True)
         self.restore_dir("/etc/xdg/systemd/user", user_post_restore, True)
+        self.restore_dir("/home/admin/.config/systemd/user", user_post_restore, True)
```
Comment (martinpitt) [inline test/verify/check-system-services:48]: Mismatch with make_test_service below, which creates the file in /etc/systemd/user/. I think it's better to create it in ~admin, as it's more realistic, but the test needs to pick one consistently.

File: test/verify/check-system-services [hunk 8]
```diff
@@ -137,7 +138,7 @@ class TestServices(MachineCase):
 
     def make_test_service(self, path="/etc/systemd/system/"):
         self.write_file(f"{path}/test.service",
-                        """
+                        '''
```
Comment (martinpitt) [inline test/verify/check-system-services:141]: Let's revert these, unrelated noise.

File: test/verify/check-system-services (hunk 7 above)
Comment (martinpitt) [reply test/verify/check-system-services:48]: Ah, this is for re-dropping the enablement symlink, as the test does not explicitly disable the service. I made it a bit more generic by restoring all of ~admin/.config.

File: test/verify/check-system-services (hunk 8 above)
Comment (martinpitt) [reply test/verify/check-system-services:141]: Done.

**Comment by martinpitt** [comment]
I robustified the test case a bit. I locally ran it 10 times in a row successfully, let's see how it looks on CI.

File: pkg/systemd/logDetails.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/systemd/logDetails.jsx:60]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230123-075627-965d915b-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 3 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:61]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230123-075627-965d915b-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/systemd/logsHelpers.js (hunk 4 above)
Comment (cockpituous) [inline pkg/systemd/logsHelpers.js:64]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18032-20230123-075627-965d915b-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by martinpitt — APPROVED** [review]
I'm fine with this now, thanks @t1m0thyj and @KKoukiou for the fixes!

@jelly , I meddled with this now, can you please do the final re-review?

**Review by KKoukiou — APPROVED** [review]

**Pull request merged**